### ISSUE TITLE
Severe oplog converter issue

### ISSUE THREAD

**Issue by arggh** [meteor/meteor#12325]
We ran into a weird and quite dangerous bug yesterday. This didn't come up during local development, since we didn't use replica sets locally (we will from now on!).

*TL;DR*

Valid update operations such as ...

```js
docs.update(
      {
        _id: doc._id,
      },
      {
        $set: {
          "items.0.id": "m57DsX8g8L66bM5JX",
          "items.0.name": "John Doe",
          "items.0.bio.en": "Just John",
          "items.0.languages.0.key": "en",
          "items.0.languages.0.englishName": "English",
          "items.0.languages.0.localName": "English",
          "items.1": {
            id: "FJwSQHqwpenCN6RQH",
            name: "Bill Clinton",
            title: {
              en: "Just Bill",
              sv: "",
            },
            bio: {
              en: "Former President",
              sv: "",
            },
            avatar: null,
            languages: [
              {
                key: "sv",
                englishName: "Swedish",
                localName: "Sverige",
              },
            ],
          },
        },
      }
    );
```

...will break Meteor's oplog converter and throw:

```
Error in oplog callback TypeError: Cannot read property 'a' of null
    at isArrayOperator (packages/mongo/oplog_v2_converter.js:46:19)
    at packages/mongo/oplog_v2_converter.js:91:11
    at Array.forEach (<anonymous>)
    at convertOplogDiff (packages/mongo/oplog_v2_converter.js:71:24)
    at packages/mongo/oplog_v2_converter.js:109:9
    at Array.forEach (<anonymous>)
    at convertOplogDiff (packages/mongo/oplog_v2_converter.js:71:24)
    at packages/mongo/oplog_v2_converter.js:109:9
    at Array.forEach (<anonymous>)
    at convertOplogDiff (packages/mongo/oplog_v2_converter.js:71:24)
    at oplogV2V1Converter (packages/mongo/oplog_v2_converter.js:122:3)
    at packages/mongo/oplog_observe_driver.js:607:16
    at Object.Meteor._noYieldsAllowed (packages/meteor.js:785:12)
    at OplogObserveDriver._handleOplogEntrySteadyOrFetching (packages/mongo/oplog_observe_driver.js:579:12)
    at packages/mongo/oplog_observe_driver.js:133:20
    at packages/mongo/oplog_observe_driver.js:17:9
    at Object.Meteor._noYieldsAllowed (packages/meteor.js:785:12)
    at packages/mongo/oplog_observe_driver.js:121:16
    at packages/mongo/oplog_tailing.js:108:7
    at runWithEnvironment (packages/meteor.js:1347:24)
    at Object.callback (packages/meteor.js:1360:14)
    at packages/ddp-server/crossbar.js:114:36
```

...then:

→ client will not receive the updates made via pub/sub, **nor any errors**.
→ client will have invalid & stale data
→ client will initiate invalid & destructive write operations based on that bad data.

_Severity:_ We already know of cases where users have lost data due to to this bug, and are losing / corrupting more data as I type this.

_Sidenote:_ At least @lukejagodzinski 's `jagi:astronomy` produces (perfectly valid) db operations that currently break Meteor's oplog converter.

**Reproduction**: https://github.com/arggh/meteor-oplog-converter-issue

Confirmed with Meteors 2.6.1 ... 2.8.1 and MongoDB 5.06 ... 5.0.14 (Mac OS, Ubuntu)

**Comment by radekmie** [comment]
Thanks for the report @arggh! I'll take it from here.

**Comment by arggh** [comment]
Great! I don't mean to rush, but do you expect to ship a fix within the next 48 hours? (If no, then I'm requested to start patching this myself as it's quite urgent for us)

**Comment by radekmie** [comment]
I can't give you an ETA, but I'll _most likely_ find time to start it in that time. If you'd like to check it yourself, then [here's the `oplogV2V1Converter` implementation](https://github.com/meteor/meteor/blob/devel/packages/mongo/oplog_v2_converter.js), and start with creating a test case in [here](https://github.com/meteor/meteor/blob/devel/packages/mongo/oplog_v2_converter_tests.js) (or multiple ones, preferably).

**Comment by arggh** [comment]
Fixed in #12339
